File: leetcode/src/00417_pacific_atlantic_water_flow/solution.rs
use std::collections::HashSet;
use std::collections::VecDeque;
use std::iter::FromIterator;

struct Solution;

impl Solution {
    pub fn pacific_atlantic(heights: Vec<Vec<i32>>) -> Vec<Vec<i32>> {
        let m = heights.len() as i32;
        let n = heights[0].len() as i32;
        let pacific = (0..m)
            .map(|i| (i, 0))
            .chain((0..n).map(|j| (0, j)))
            .collect::<Vec<_>>();
        let atlantic = (0..m)
            .map(|i| (i, n - 1))
            .chain((0..n).map(|j| (m - 1, j)))
            .collect::<Vec<_>>();
        Solution::bfs(&heights, m, n, &pacific)
            .intersection(&Solution::bfs(&heights, m, n, &atlantic))
            .map(|&x| vec![x.0, x.1])
            .collect()
    }

    fn bfs(heights: &[Vec<i32>], m: i32, n: i32, starts: &[(i32, i32)]) -> HashSet<(i32, i32)> {
        let dirs: &'static [i32] = &[0, 1, 0, -1, 0];
        let mut q = VecDeque::from_iter(starts.iter().cloned());
        let mut visited = HashSet::from_iter(starts.iter().cloned());
        while let Some((ci, cj)) = q.pop_back() {
            for i in 0..4 {
                let (ni, nj) = (ci + dirs[i], cj + dirs[i + 1]);
                if ni >= m
                    || ni < 0
                    || nj >= n
                    || nj < 0
                    || visited.contains(&(ni, nj))
                    || heights[ni as usize][nj as usize] < heights[ci as usize][cj as usize]
                {
                    continue;
                }
                visited.insert((ni, nj));
                q.push_back((ni, nj));
            }
        }
        visited
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test() {
        let heights = vec![
            vec![1, 2, 2, 3, 5],
            vec![3, 2, 3, 4, 4],
            vec![2, 4, 5, 3, 1],
            vec![6, 7, 1, 4, 5],
            vec![5, 1, 1, 2, 4],
        ];
        assert_eq!(
            Solution::pacific_atlantic(heights)
                .into_iter()
                .collect::<HashSet<Vec<i32>>>(),
            vec![
                vec![0, 4],
                vec![1, 3],
                vec![1, 4],
                vec![2, 2],
                vec![3, 0],
                vec![3, 1],
                vec![4, 0]
            ]
            .into_iter()
            .collect()
        );

        let heights = vec![vec![1]];
        assert_eq!(Solution::pacific_atlantic(heights), vec![vec![0, 0]]);
    }
}

File: leetcode/src/00417_pacific_atlantic_water_flow/solution.go
package leetcode

type Point struct {
	x int
	y int
}

func pacificAtlantic(heights [][]int) [][]int {
	m := len(heights)
	n := len(heights[0])
	pacific := []Point{}
	atlantic := []Point{}
	for i := 0; i < m; i++ {
		pacific = append(pacific, Point{x: i, y: 0})
		atlantic = append(atlantic, Point{x: i, y: n - 1})
	}
	for j := 0; j < n; j++ {
		pacific = append(pacific, Point{x: 0, y: j})
		atlantic = append(atlantic, Point{x: m - 1, y: j})
	}
	var res [][]int
	visited1 := bfs(heights, m, n, pacific)
	visited2 := bfs(heights, m, n, atlantic)
	for p := range visited1 {
		if _, exist := visited2[p]; exist {
			res = append(res, []int{p.x, p.y})
		}
	}
	return res
}

func bfs(heights [][]int, m int, n int, starts []Point) map[Point]struct{} {
	dirs := [5]int{-1, 0, 1, 0, -1}
	q := append([]Point(nil), starts...)
	visited := make(map[Point]struct{}, len(starts))
	for _, p := range starts {
		visited[p] = struct{}{}
	}
	for len(q) > 0 {
		var p Point
		p, q = q[0], q[1:]
		for i := 0; i < 4; i++ {
			np := Point{
				x: p.x + dirs[i],
				y: p.y + dirs[i+1],
			}
			if np.x >= m || np.x < 0 || np.y >= n || np.y < 0 {
				continue
			}
			if _, exist := visited[np]; exist {
				continue
			}
			if heights[np.x][np.y] < heights[p.x][p.y] {
				continue
			}
			q = append(q, np)
			visited[np] = struct{}{}
		}
	}
	return visited
}
